websocket-chat-broker/src/main.rs

```diff
@@ -0,0 +1,156 @@
+#[macro_use]
+extern crate actix;
+extern crate actix_web;
+extern crate actix_broker;
+extern crate futures;
+extern crate rand;
+
+use actix::prelude::*;
+use actix::fut;
+use actix_web::server::HttpServer;
+use actix_web::{fs, http, ws, App, Error, HttpRequest, HttpResponse};
+use actix_broker::BrokerIssue;
+use futures::Future;
+
+mod server;
+use server::*;
+
+fn chat_route(req: &HttpRequest<()>) -> Result<HttpResponse, Error> {
+    ws::start(req, WsChatSession::default())
+}
+
+#[derive(Default)]
+struct WsChatSession {
+    id: usize,
+    room: String,
+    name: Option<String>,
+}
+
+impl WsChatSession {
+    fn join_room(&mut self, room_name: &str, ctx: &mut ws::WebsocketContext<Self>) {
+        let room_name = room_name.to_owned();
+        // First send a leave message for the current room
+        let leave_msg = LeaveRoom(self.room.clone(), self.id);
+        self.issue_sync(leave_msg, ctx);
+        // Then send a join message for the new room
+        let join_msg = JoinRoom(
+            room_name.to_owned(),
+            self.name.clone(),
+            ctx.address().recipient());
+
+        WsChatServer::from_registry()
+            .send(join_msg)
+            .map_err(|_| ())
+            .into_actor(self)
+            .map(|id, act, _| {
+                act.id = id;
+                act.room = room_name;
+            })
+            .wait(ctx);
+    }
+
+    fn list_rooms(&mut self, ctx: &mut ws::WebsocketContext<Self>) {
+        WsChatServer::from_registry()
+            .send(ListRooms)
+            .into_actor(self)
+            .then(|res, _, ctx| {
+                if let Ok(rooms) = res {
+                    for room in rooms {
+                        ctx.text(room);
+                    }
+                }
+                fut::ok(())
+            }).wait(ctx);
+    }
+
+    fn send_msg(&self, msg: &str) {
+        let content = format!("{}: {}",
+                              self.name.clone().unwrap_or("anon".to_string()),
+                              msg);
+        let msg = SendMessage(self.room.clone(), self.id, content);
+        self.issue_async(msg);
+    }
+}
+
+impl Actor for WsChatSession {
+    type Context = ws::WebsocketContext<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        self.join_room("Main", ctx);
+    }
+
+    fn stopped(&mut self, _ctx: &mut Self::Context) {
+        println!("WsChatSession stopped for {}({}) in room {}",
+            self.name.clone().unwrap_or("anon".to_string()),
+            self.id,
+            self.room);
+    }
+}
+
+impl Handler<ChatMessage> for WsChatSession {
+    type Result = (); 
+
+    fn handle(&mut self, msg: ChatMessage, ctx: &mut Self::Context) {
+        ctx.text(msg.0);
+    }
+}
+
+impl StreamHandler<ws::Message, ws::ProtocolError> for WsChatSession {
+    fn handle(&mut self, msg: ws::Message, ctx: &mut Self::Context) {
+        println!("WEBSOCKET MESSAGE: {:?}", msg);
```

Comment:
debug println, maybe to use logging in this example?

Reply:
Added `log` and `simple_logger` crates. Changed all `println!`'s.